accepted/2021/System.Diagnostics/Metrics-Design.md

```diff
@@ -0,0 +1,502 @@
+    }
+}
+```
+
+### Observable Instrument Base class
+
+```csharp
+namespace System.Diagnostics.Metrics
+{
+    /// <summary>
+    /// ObservableInstrument<T> is the base class from which all observable instruments will inherit from.
+    /// Mainly It'll support the CLS compliant numerical types
+    /// </summary>
+    public abstract class ObservableInstrument<T> : Instrument where T : unmanaged
+    {
+        /// <summary>
+        /// Protected constructor to create the instrument with the common properties.
+        /// </summary>
+        protected ObservableInstrument(
+                    Meter meter,
+                    string name,
+                    string? description,
+                    string? unit) : base(meter, name, description, unit) { throw null; }
+
+        /// <summary>
+        /// Observe will report the non-additive values of the measurements.
+        /// </summary>
+        protected abstract IEnumerable<Measurement<T>> Observe();
+    }
+}
+```
+
+### Measurement class
+
+```csharp
+
+namespace System.Diagnostics.Metrics
+{
+    /// <summary>
+    /// A helper class used by the Observable instruments Observe method to repro the measurement values
+    /// with the associated tags.
+    /// </summary>
+    public struct Measurement<T> where T : unmanaged
+    {
+        /// <summary>
+        /// Construct the Measurement using the value and the list of tags.
+        /// We'll always copy the input list as this is not perf hot path.
+        /// </summary>
+        public Measurement(T value, IEnumerable<KeyValuePair<string, object>> tags) { throw null; }
+        public Measurement(T value, params KeyValuePair<string, object?>[] tags) { throw null; }
+
+        public ReadOnlySpan<KeyValuePair<string, object?>> Tags { get { throw null;  } }
+        public T Value { get; }
+    }
+}
+```
+
+### Non-observable Instruments
+
+```csharp
+namespace System.Diagnostics.Metrics
+{
+    /// <summary>
+    /// The counter is a non-observable Instrument that supports non-negative increments.
+    /// e.g. Number of completed requests.
+    /// </summary>
+    public class Counter<T> : Instrument<T> where T : unmanaged
+    {
+        public void Add(T measurement) { throw null; }
+        public void Add(T measurement,
+                            KeyValuePair<string, object> tag1) { throw null; }
+        public void Add(T measurement,
+                            KeyValuePair<string, object> tag1,
+                            KeyValuePair<string, object> tag2) { throw null; }
+        public void Add(T measurement,
+                            KeyValuePair<string, object> tag1,
+                            KeyValuePair<string, object> tag2,
+                            KeyValuePair<string, object> tag3) { throw null; }
+        public void Add(T measurement,
+                            params KeyValuePair<string, object>[] tags) { throw null; }
```

Comment:
Do we want to take a ReadOnlySpan<KeyValuePair<string,Object>> as well? Since we return ReadOnlySpan in the Measurment.Tags

Reply:
I think this would be a good idea. at least can allow passing a range of items inside the array. I'll add that overload.

Reply:
Just to tell, I'll add a constructor overload which take ReadOnlySpan to the `Measurement<T>` struct.